### DISCUSSION TITLE
Tarik data from ODBC 2

### DISCUSSION THREAD

**Discussion by joejnt** [teamalgoritma/community#128] [Q&A] [answered]
Dear Tim Algoritma,

Mohon bantuannya, saat saya tarik data dari database (ODBC), untuk beberapa data hasilnya sukses. Namun pada saat tarik data ke 15 dan berikutnya selalu terjadi error. Untuk syntax yang saya gunakan adalah sebagai berikut:

Untuk tabel 18
db_zc0018 <- dbConnect(
  MySQL(),
  port = xxx,
  host = "xx.xxx.x.xx",
  user = "skai",
  password = "xxxx",
  dbname = "database"
)

tablezc0018 <- dbSendQuery(conn = db_zc0018, statement = "SELECT * FROM ZC0018")
ZC0018 <- dbFetch(tablezc0018)
ZC0018

Untuk tabel 19
db_zc0019 <- dbConnect(
  MySQL(),
  port = xxx,
  host = "xx.xxx.x.xx",
  user = "skai",
  password = "xxxx",
  dbname = "database"
)

tablezc0019 <- dbSendQuery(conn = db_zc0019, statement = "SELECT * FROM ZC0019")
ZC0019 <- dbFetch(tablezc0019)
ZC0019

keterangan error: 
Error in .local(drv, ...) : 
Cannot allocate a new connection: 16 connections already opened

Adapun untuk data ZC0018 total observasi sebanyak 511.000 dengan total variabel 190. Total data yang akan ditarik sebanyak 30 dataset. Apakah karena data nya yang terlalu besar sehingga terjadi error.

Mohon saran dan petunjuknya.
Terima kasih,

Best regards,
JT

**Comment by inytss** [comment, marked as answer]
Dear @joejnt ,
Sepertinya Bapak terlalu banyak membuka koneksi pada database yang sama. Ketika Bapak ingin melakukan fetching data masih dengan database yang sama, Bapak cukup menggunakan satu koneksi saya, misalnya dengan connection Bapak saat mengambil tabel pertama. Connection tersebut dapat digunakan berulang kali untuk menarik tabel dalam satu database yang sama. 
Sebagai solusi, Bapak dapat menonaktifkan connection terhadap databasenya terlebih dahulu dan mencobanya kembali seperti pada solution yang ditawarkan pada laman berikut https://stackoverflow.com/questions/32139596/cannot-allocate-a-new-connection-16-connections-already-opened-rmysql. Atau Bapak bisa mencoba untuk melakukan terminate RStudionya agar koneksi yang pernah dibuat terlepas semua.

Semoga jawaban ini cukup membantu ya. Jika Bapak merasa jawaban ini membantu, Bapka dapat menekan tombol "_Mark as answer_".

**Comment by joejnt** [comment]
Dear @inytss 

Terima kasih untuk insightnya, namun saya masih bingung dengan artikel yang diberikan. Jadi ketika saya jalankan koneksi dengan script sebagai berikut:

# Data ZC0001
db_zc0001 <- dbConnect(
  MySQL(),
  port = 330x,
  host = "10.xxx.2.xx",
  user = "user",
  password = "mypassword",
  dbname = "DB_AOS"
)

tablezc0001 <- dbSendQuery(conn = db_zc0001, statement = "SELECT * FROM ZC0001")
ZC0001 <- dbFetch(tablezc0001)
ZC0001

kemudian saya terminate koneksinya dengan menjalankan script ini:

sqlQuery <- function (query) {
  # creating DB connection object with RMysql package
  DB <- dbConnect(MySQL(), 
                  port = 330x,
                  user="user", 
                  password='mypassword', 
                  dbname='DB_AOS', 
                  host='10.xxx.2.xx')
  # close db connection after function call exits
  on.exit(dbDisconnect(DB))
  # send Query to btain result set
  rs <- dbSendQuery(DB, query)
  # get elements from result sets and convert to dataframe
  result <- fetch(rs, -1)
  # return the dataframe
  return(result)
}


Kemudian saya membuat koneksi baru dengan menjalankan script:
db_zc0002 <- dbConnect(
  MySQL(),
  port = 330x,
  host = "10.xxx.2.xx",
  user = "user",
  password = "mypassword",
  dbname = "DB_AOS"
)

tablezc0002 <- dbSendQuery(conn = db_zc0002, statement = "SELECT * FROM ZC0002")
ZC0002 <- dbFetch(tablezc0002)
ZC0002

Begitu dan seterusnya ya ?
Mohon koreksinya.

Best regards,
JT

**Reply by joejnt** [reply]
Dear @inytss 

Untuk permasalahan diatas sudah selesai dengan saya melakukan terminate connection saat selesai melakukan query data. Script yang saya gunakan adalah sebagai berikut:

db <- dbConnect(
  MySQL(),
  port = 3xxx,
  host = "10.xxx.x.xx",
  user = "user",
  password = "password",
  dbname = "DB_AOS"
)

tablezf0006 <- dbSendQuery(conn = db, statement = "SELECT * FROM ZF0006")
ZF0006 <- dbFetch(tablezf0006)
dbDisconnect(db)
ZF0006

Saya mendapatkan permasalahan, pada data yang saya query dari total observasi sebenarnya ada 531.117.  Namun setelah saya query menggunakan R, total observasinya hanya 500. Saya memastikan lagi dengan mendownload data dalam bentuk 'xlsx', dan benar total observasi hanya sampai 500 saja.

Apakah kondisi tersebut terjadi karena ukurannya yang terlalu besar ?

mohon saran dan petunjuknya.

Best regards,
JT

**Reply by inytss** [reply]
Dear @joejnt ,
Mengenai hal tersebut boleh dipastikan kembali pada databasenya apakah memang terdapat 531.117 data. Karena untuk data sendiri seharusnya tidak terbatas berdasarkan jumlah observasinya pak.
